libpod/runtime_img.go

```diff
@@ -76,6 +80,79 @@ type ImageFilterParams struct {
 	ImageInput       string
 }
 
+// struct for when a user passes a short or incomplete
+// image name
+type imagePullStruct struct {
+	imageName   string
+	tag         string
+	registry    string
+	hasRegistry bool
+	transport   string
+}
+
+func (ips imagePullStruct) returnFQName() string {
+	return fmt.Sprintf("%s%s/%s:%s", ips.transport, ips.registry, ips.imageName, ips.tag)
+}
+
+func getRegistriesToTry(image string) ([]string, error) {
+	var registries []string
+	var imageError = fmt.Sprintf("unable to parse '%s'\n", image)
+	imgRef, err := reference.Parse(image)
+	if err != nil {
+		return nil, errors.Wrapf(err, imageError)
+	}
+	tagged, isTagged := imgRef.(reference.NamedTagged)
+	tag := "latest"
+	if isTagged {
+		tag = tagged.Tag()
+	}
+	hasDomain := true
+	registry := reference.Domain(imgRef.(reference.Named))
+	if registry == "" {
+		hasDomain = false
+	}
+	imageName := reference.Path(imgRef.(reference.Named))
+	pImage := imagePullStruct{
+		imageName,
+		tag,
+		registry,
+		hasDomain,
+		"docker://",
+	}
+	if pImage.hasRegistry {
+		// If input has a registry, we have to assume they included an image
+		// name but maybe not a tag
+		_, err := alltransports.ParseImageName(pImage.returnFQName())
```

Comment:
This seems like a pretty complex way of writing, essentially, `docker.ParseReference(image)`, on this side of the `if`.

(It’s not _quite_ the same, because `returnFQName` discards the digest, if any; but I guess that the intent was to preserve the original semantics.  (`docker.ParseReference` is also calling `reference.TagNameOnly`, so that doesn’t need to be done by this function either.))